### ISSUE TITLE
Draggable balloon not released on focus change from Virtual Cursor

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/balloons-and-static-electricity#187]
In #183, @terracoda said

> @jessegreenberg, another complication is that the grabbed state is easily lost visually by using the cursor keys to read the PDOM, but the balloon doesn't release when the user does that, and then it is difficult to grab it again because it is already grabbed.

@terracoda sorry!  I am happy to see that the cursor keys work well with VoiceOver when focus is on an application element!  The balloon will be released when the draggable balloon is no longer the active element.

This needs to be fixed immediately.

**Comment by jessegreenberg** [comment]
Unfortunately, VoiceOver does not fire ```blur``` when focus changes with the Virtual Cursor.

**Comment by jessegreenberg** [comment]
Bah...I was testing in a build...

**Comment by jessegreenberg** [comment]
Releasing the balloon works, but need to verify that adding this doesn't break other features.

**Comment by jessegreenberg** [comment]
The fix seems to break NVDA + Firefox.

**Comment by jessegreenberg** [comment]
I wonder if NVDA blurs the balloon for a second when a key is pressed?

**Comment by jessegreenberg** [comment]
As a temporary work around, we can try to release the balloon if it is picked up and the Remove Wall button has focus.

**Comment by jessegreenberg** [comment]
@terracoda temporary workaround is to focus release balloon when the remove wall button is focused.  In general, we do not know where the VirtualCursor is so we cannot release the balloon when the user reads the Wall description (for instance).  It would be good to discuss the example where this was confusing for the user.

**Comment by jessegreenberg** [comment]
Removing top priority since it will require discussion before fixing.

**Comment by terracoda** [comment]
@jessegreenberg, I think the main issue is that with no interactive alerts firing the user has no feedback from interaction, so they go the PDOM to read what is going on. When they do that, the actual visual focus disappeared, so visually I could not see where their keyboard focus was. I also could not see if the balloon was grabbed or released. The description for the balloon read out  as if it were grabbed (and it was grabbed), but the WASD keys did not work until the user found a way to get the visual keyboard focus back on the balloon.

I am going to test the new version now to see how it works :-)

**Comment by terracoda** [comment]
@jessegreenberg, FYI, 3 interaction alerts just read out when i was not expecting them! This makes me think there is some serious delay issues happening. This happened while using the web version. I will try a new downloaded version to see if that works better.

**Comment by jessegreenberg** [comment]
@terracoda that is probably related to #178, what browser are you using?

**Comment by terracoda** [comment]
@jessegreenberg, the new downloaded version does not read out any interaction alerts, not even the simple ones for Add/Remove Wall, Two-balloon experiment, or the reset buttons. These all worked when I first tried the sim on Friday. Some alerts were a little intermittent, not always working. 

Also noticed that keyboard focus state does change (though visual focus disappears) when I arrow to new interactive elements like the Wall button, but the balloon was not been released, even when I removed the  wall. When I arrowed back to the balloon, when the visual focus returned, the outline was black indicating a grabbed state. There is something not working, not sure how to describe it.

**Comment by terracoda** [comment]
@jessegreenberg, I am only using Chrome and VO right now as according to your testing that was the best combo - avoiding Safari right now :-)

**Comment by terracoda** [comment]
@jessegreenberg my original use on Friday was also with Chrome and VO.

**Comment by terracoda** [comment]
@jessegreenberg, just noticed that with the downloaded version I can get stick on the balloon, i.e. I can't tab away from it. This does not happen on the web version.

**Comment by jessegreenberg** [comment]
> I think the main issue is that with no interactive alerts firing the user has no feedback from interaction,

I see, so maybe we should wait to work on this until we see it is an issue on a system where all features are working correctly.

**Comment by jessegreenberg** [comment]
> @jessegreenberg, just noticed that with the downloaded version I can get stick on the balloon, i.e. I can't tab away from it. This does not happen on the web version.

@terracoda I cannot reproduce this issue on any of my devices.  If this happens again can you please open a new github issue with details?

**Comment by terracoda** [comment]
@jessegreenberg, yes, if this happens again I will open an issue. Once the downloaded version is back up to speed, we can test for this.

**Comment by jessegreenberg** [comment]
@terracoda when the alerts are working correctly, is this still an issue?

**Comment by jessegreenberg** [comment]
11/2/16 meeting - @terracoda will reproduce this issue on a platform where aria-live alerts are working correctly.  If the issue can be reproduced, please add a comment with steps on how to reproduce the bug and information about what platform was used.

**Comment by jessegreenberg** [comment]
@terracoda also please reassign me after you have investigated :)

**Comment by jessegreenberg** [comment]
In #193 @terracoda said 
> After grabbing balloon, I can use CO-Right or Left arrow to read elsewhere, after to arrowing back to the balloon the black Grab focus is still there, but I cannot use the WASD keys to move it. I need to hit Space bar of Enter Key again to activate Application again (interesting). (Maybe we can address this in a the help documentation)

This does not happen in NVDA or JAWS (arrow keys do *nothing* on the application role), so I will have to investigate tomorrow when I have access to an OS device.

Also, removing high priority label since this is an OSX/MacOS issue and next interviews will be conducted with NVDA.

**Comment by terracoda** [comment]
@jessegreenberg, good to know this is an issue on MacOS. I think the issue exist also when using ChromeVox on MacOS. We can investigate further after next week's interviews.

**Comment by jessegreenberg** [comment]
> I think the issue exist also when using ChromeVox on MacOS.

Good point!  If an isssue on ChromeVox, I would expect it to happen using ChromeVox on Windows as well, I can check.

**Comment by jessegreenberg** [comment]
Hmm, doesn't seem to happen with ChromeVox + Windows.  The arrow keys do nothing.  I suspect that we would be able to use the arrow keys with role=application and ChromeVox.

**Comment by jessegreenberg** [comment]
I was able to reproduce this on a MacBook Air with Yosemite.  The main problem is that once the virtual cursor leaves the balloon, it cannot find it again because the draggable balloon looks like this:
```html
<div role=application ...(other attributes)></div>
```

So when the focus hits the balloon again, it only finds the 'Grab Balloon' button, which is why dragging doesn't work.

To fix this issue, the div above will need some content for the virtual cursor to find.  It would also be best to hide the 'grab' button to avoid confusion.

**Comment by jessegreenberg** [comment]
Here is the html representing the entire balloon widget:
```html
<div id="518">
	<h3 id="label-518">‪Yellow Balloon‬</h3>
    <p id="description-518">In upper center of play area. Has net neutral charge, no more negative charges than positive charges. Look for grab button to play.</p>
	<div>
		<div>
			<button id="590">‪Grab ‪Yellow Balloon‬‬</button>
			<p id="description-590">Once grabbed, get ready to drag balloon.  Press W, A, S, or D key to drag up, left, down, or right.  To let go, press Space bar.</p>
	    </div>
		<div>
		  <div id="589" role="application" tabindex="-1" aria-describedby="description-518" aria-labelledby="label-518"></div>
       </div>
	</div>
</div>
```

I am wondering if this could be fixed by changing to the following:
```html
<div id="518">
	<h3 id="label-518">‪Yellow Balloon‬</h3><p id="description-518">In upper center of play area. Has net neutral charge, no more negative charges than positive charges. Look for grab button to play.</p>
  <div>
    <div id='button-container'>
      <button id="590">‪Grab ‪Yellow Balloon‬‬</button>
      <p id="description-590">Once grabbed, get ready to drag balloon.  Press W, A, S, or D key to drag up, left, down, or right.  To let go, press Space bar.</p>
    </div>
    <div id='application-container' hidden>
        <div id="589" role="application" tabindex="-1" aria-describedby="description-518">Yellow Balloon</div>
    </div>
  </div>
</div>
```
This might work if we hide the button container div and unhide the application container div when the balloon is grabbed, do the opposite when the balloon is released.  We would probably remove the aria-labelledby attribute from the draggable balloon, and add the label as inner content to the div with role='application'.

The problem with aria-labelledby in this context is that it is only ever read when the balloon receives programmatic focus.  Once focus leaves the div, it will never go back to the application div unless the user presses 'tab'.

**Comment by jessegreenberg** [comment]
Another benefit is that the 'Once grabbed, get...' cue will be hidden once the balloon is picked up.  This will need to be hidden if VoiceOver can use the virtual cursor while the balloon is picked up.

**Comment by terracoda** [comment]
@jessegreenberg,  Yes, I think this solution makes better sense. 

On a semi-related note, the first solution explains why I was hearing "Yellow Balloon" twice sometimes. Once as the h3 and once as the the label for the empty application div. I thought it was because I was leaving a region, but applications no longer have the default region role, so I was confused. In fact, it was because the label for the empty application div was being read out.

I didn't realize the application div was a separate div on its own. I imagined that the heading and the description become enveloped in application div, once grabbed, i.e. the content elements together became the application widget. Maybe that is not possible, in which case, I think your new structure is indeed better than the first.

Structuring these interactive widgets is an interesting design problem. On traditional websites, I generally like to use way fewer divs, but maybe that's not possible in the case of these scenarios.

**Comment by jessegreenberg** [comment]
Ok, thanks @terracoda, I will give this a shot.  For enveloping the whole thing in the application role, we actually used to do this, but I decided against it after coming across this statement from the ARIA roles model docs:

> Roles are element types and authors MUST NOT change role values over time or with user actions. Authors wishing to change a role MUST do so by deleting the associated element and its children and replacing it with a new element with the appropriate role.

Deleting the whole element and building it up again would be trickier, and placing the application role on the whole thing would prevent the user from being able to use the virtual cursor to read description info.

As for the number of divs, placing groups of elements under a parent div makes it easier for scenery to place things in the right location in the PDOM and also easier to determine which visual element should receive the focus highlight.

**Comment by terracoda** [comment]
@jessegreenberg thanks for explaining. Sorry, I forgot about the change. Many frameworks work with many divs. I do understand their non-semantic usefulness of wrapping things together.

**Comment by jessegreenberg** [comment]
This is a pretty big change to the interaction, it would be great if it were included in the next interview.  It is also pretty intrusive, we will have to be sure to test it out.

**Comment by jessegreenberg** [comment]
Right now, the implementation for setHidden in AccessibleNode simply sets the Node's domElement to hidden.  As illustrated in https://github.com/phetsims/balloons-and-static-electricity/issues/187#issuecomment-258560918, we will actually want to hide the parent container element so that the element, label, and descriptions (and any other html that might be under the element) are also hidden.

**Comment by terracoda** [comment]
@jessegreenberg, we are doing this for next, after testing, right?

**Comment by jessegreenberg** [comment]
Either way, It is sort of a structural change to the balloon widget, so I thought it would be nice to include in this round of testing.  But if you would rather wait, that is fine too.

**Comment by jessegreenberg** [comment]
On focus of 
```html
<div id="589" role="application" tabindex="-1" aria-describedby="description-518">Yellow Balloon</div>
```
NVDA does not read 'Yellow Balloon' on focus.  It only reads 'Application' and then the aria-describedby text.  So we will likely need a combination of aria-labelledby and inner text so that the virtual cursor has something to read AND the label is read immediately.

**Comment by terracoda** [comment]
@jessegreenberg, agree we will need to try with aria-labelledyby.

**Comment by jessegreenberg** [comment]
https://github.com/phetsims/balloons-and-static-electricity/issues/187#issuecomment-258560918 fixes most of the problem.  When the user navigates away from the balloon with the virtual cursor, the user no longer has to 'pick the balloon up again' to interact with it.

Next part of the problem is that the balloon isn't released if the user leaves the balloon by using the virtual cursor.  @terracoda any thoughts on how to solve this problem?

One option would be to drop the balloon when the next or previous form element receives focus. Would this be OK?  The balloon would still have focus (and be picked up) while the user is reading content that isn't focusable (like the wall description, for instance).

**Comment by terracoda** [comment]
@jessegreenberg, yes, I think focus on the next control is the best way to handle this issue.

**Comment by jessegreenberg** [comment]
Sounds good @terracoda, I will work on that.

**Comment by jessegreenberg** [comment]
Since the arrow keys now move the balloon and don't navigate away, this is no longer an issue. Closing.
